### PR TITLE
[Storage][DataMovement] Add bounds to transfer queues

### PR DESCRIPTION
**Pull request by jalauzon-msft**
This change adds bounds to the various channels in an attempt to limit memory usage. The focus for this change was on download.

- Change `ChannelProcessing` to be able to create bounded channels and fix bug where number of readers was being hardcoded rather than using configured value. Also switch `DownloadChunkHandler` to use `ChannelProcessing`.
- Reduce parts readers to 32 and add capacity to download chunk channel of 16. At 4 MiB chunks, this roughly equates to 2G of memory as a worst case (working on 32 files, each with 16 items in queue). This is not an exact max as additional files could have been started.
- Change default `MaxConcurrency ` (chunk channel reader count) to `Environment.ProcessorCount * 8` to match Track1. Reduced from 3000 which was too high and would overwhelm most systems.
- Add capacity of 1000 to part and chunk channels.
  - While things in these channels do not consume much memory, the processing for filling these channels can be extremely quick and so these queues could fill indefinitely until the entire job was parsed. Setting a capacity means we won't have runaway queue depth in very large transfers and also allows for processing time to spent elsewhere once these channels fill. 1000 is value high enough to never really affect performance.
  - This change is what I believe has resulted in significantly better performance (>10x) on copies with many small files.

Overall, this change has not hurt, and in some cases significantly improved, performance while somewhat bounding the potential memory usage during download.

### CODE REVIEW

**Comment by jalauzon-msft** [comment]
Perf results - https://dev.azure.com/azure-sdk/internal/_build/results?buildId=4366820&view=logs&j=42ded549-05ee-5cdb-7ba7-7a948a0cc056&t=1a18f1cb-d016-5e13-bec4-a5dc2ac35679

**Comment by jalauzon-msft** [comment]
@microsoft-github-policy-service rerun

File: sdk/storage/Azure.Storage.DataMovement/src/Shared/DataMovementConstants.cs [hunk 1]
```diff
@@ -1,26 +1,27 @@
...
         internal const long DefaultChunkSize = 4 * Constants.MB;
 
         public const char PathForwardSlashDelimiterChar = '/';
 
+        internal static class Channels
+        {
+            internal const int MaxJobPartReaders = 32;
+            internal static int MaxJobChunkReaders = Environment.ProcessorCount * 8;
+            internal const int JobPartCapacity = 1000;
+            internal const int JobChunkCapactiy = 1000;
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement/src/Shared/DataMovementConstants.cs:21]: NIT - Capacity*

File: sdk/storage/Azure.Storage.DataMovement/src/ChannelProcessing.cs [hunk 2]
```diff
@@ -15,27 +15,32 @@ namespace Azure.Storage.DataMovement;
...
 {
     ValueTask QueueAsync(TItem item, CancellationToken cancellationToken = default);
+    bool TryComplete();
     ProcessAsync<TItem> Process { get; set; }
 }
 
 internal static class ChannelProcessing
 {
-    public static IProcessor<T> NewProcessor<T>(int parallelism)
+    public static IProcessor<T> NewProcessor<T>(int readers, int capactiy = -1)
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement/src/ChannelProcessing.cs:24]: NIT - Capacity

**Review by amnguye — APPROVED** [review]

**Review by jaschrep-msft — APPROVED** [review]

**Comment by jalauzon-msft** [comment]
@microsoft-github-policy-service rerun

**Pull request merged**